beacon_node/network/src/sync/block_sidecar_coupling.rs

```diff
@@ -265,6 +406,70 @@ impl<E: EthSpec> RangeBlockComponentsRequest<E> {
                     }
                 }
 
+                Some(resp)
+            }
+            // Reuse same logic that we use for coupling data columns for now.
+            // todo(pawan): we should never get a coupling error here, so simplify this
+            // variant's handling.
+            RangeBlockDataRequest::DataColumnsFromRoot {
+                all_requests_made,
+                attempt,
+                column_peers,
+                expected_custody_columns,
+                requests,
+            } => {
+                // Do not couple until requests covering all required columns
+                // have been made
+                if !*all_requests_made {
+                    return None;
+                }
+
+                let mut data_columns = vec![];
+                let mut column_to_peer_id: HashMap<u64, PeerId> = HashMap::new();
+                for req in requests.values() {
+                    let Some(data) = req.to_finished() else {
+                        return None;
+                    };
+                    data_columns.extend(data.clone())
+                }
+
+                // An "attempt" is complete here after we have received a response for all the
+                // requests we made. i.e. `req.to_finished()` returns Some for all requests.
+                *attempt += 1;
+
+                // Note: this assumes that only 1 peer is responsible for a column
+                // with a batch.
+                for (id, columns) in column_peers {
```

Comment:
we need to make sure the `column_peers` map is cleaned up properly, otherwise we could end up attributing fault to the wrong peer, comment here:
https://github.com/sigp/lighthouse/pull/7946#discussion_r2324195265